examples/linear_deformation.py

```diff
@@ -0,0 +1,1187 @@
+
+        Parameters
+        ----------
+        par_space : `ProductSpace` or `Rn`
+            Space of the parameters. For one-dimensional deformations,
+            `Rn` can be used. Otherwise, a `ProductSpace` with ``n``
+            components is expected.
+        control_points : `TensorGrid` or `array-like`
+            The points ``x_j`` controlling the deformation. They can
+            be given either as a tensor grid or as a point array. In
+            the latter case, its shape must be ``(N, n)``, where
+            ``n`` is the dimension of the template space, and ``N``
+            the number of ``alpha_j``, i.e. the size of (each
+            component of) ``par_space``.
+        discr_space : `DiscreteSpace`
+            Space of the image grid of the template.
+        kernel : `callable`
+            Function to determine the kernel at the control points ``K(y_j)``
+            The function must accept a real variable and return a real number.
+        """
+        # TODO: use kernel operator instead of function & matrix
+        if isinstance(par_space, odl.Fn):
+            # Make a product space with one component
+            par_space = odl.ProductSpace(par_space, 1)
+        elif isinstance(par_space, odl.ProductSpace):
+            pass
+        else:
+            raise TypeError('expected Rn or ProductSpace as par_space, got '
+                            '{!r}.'.format(par_space))
+
+        if par_space.size != discr_space.ndim:
+            raise ValueError('dimensions of product space and image grid space'
+                             ' do not match ({} != {})'
+                             ''.format(par_space.size, discr_space.ndim))
+
+        # The operator maps from the parameter space to an inverse
+        # displacement for the domain of the target.
+
+        self.discr_space = discr_space
+        self.range_space = odl.ProductSpace(self.discr_space,
+                                            self.discr_space.ndim)
+
+        super().__init__(par_space, self.range_space, linear=True)
+
+        self.ft_kernel = ft_kernel
+
+        if not isinstance(control_points, odl.TensorGrid):
+            self._control_pts = np.asarray(control_points)
+            if self._control_pts.shape != (self.num_contr_pts, self.ndim):
+                raise ValueError(
+                    'expected control point array of shape {}, got {}.'
+                    ''.format((self.num_contr_pts, self.ndim),
+                              self.control_points.shape))
+        else:
+            self._control_pts = control_points
+
+        # TODO: check that number of control points is the same as alphas
+
+    @property
+    def ndim(self):
+        """Number of dimensions of the deformation."""
+        return self.domain.size
+
+    @property
+    def contr_pts_is_grid(self):
+        """`True` if the control points are given as a grid."""
+        return isinstance(self.control_points, odl.TensorGrid)
```

Comment:
I remember that hack. It is ugly in the sense that you need to ask "is it a grid or not" each time you want to do something with the control points. Probably a sign that we should have two different operators, one with control points on a grid, one with scattered points.